der/src/decoder.rs

```diff
@@ -283,6 +283,13 @@ impl<'a> Decoder<'a> {
         Ok(self.bytes.ok_or(ErrorKind::Failed)?.len())
     }
 
+    /// Obtain a slice of bytes contain a complete TLV production suitable for parsing later.
+    pub fn tlv_slice(&mut self) -> Result<&'a [u8]> {
+        let header = self.peek_header()?;
+        let header_len = self.peek_header()?.encoded_len()?;
```

Comment:
This is calling `peek_header()` twice and thus parsing the header twice.

```suggestion
        let header_len = header.encoded_len()?;
```